lambda/github_api.py

```diff
@@ -0,0 +1,97 @@
+import os
+import time
+
+import requests
+
+import config
+from logger import LOG
+
+TOKEN_PREFIX = "/alert-processor/tokens/github/"
+API_ROOT = "https://api.github.com"
+ACCESS_TOKEN = None
+
+
+def set_github_access_token(token):
+    """ Get the access token once """
+    global ACCESS_TOKEN
+    ACCESS_TOKEN = token
+
+
+def get_github_access_token():
+    """ Get the access token once """
+    global ACCESS_TOKEN
+    if ACCESS_TOKEN is None:
+        access_token_param = os.environ.get("GITHUB_TOKEN")
+        ACCESS_TOKEN = config.get_ssm_param(access_token_param)
+    return ACCESS_TOKEN
+
+
+def get_github_api_paged_data(url: str) -> list:
+    token = get_github_access_token()
+    page = 1
+    page_size = 100
+    page_items = 1
+    items = []
+    while page_items > 0:
+        page_items = 0
+        page_url = f"{url}?page={page}&per_page={page_size}"
+        headers = {"authorization": f"token {token}"}
+        response = requests.get(page_url, headers=headers).json()
+        if isinstance(response, list):
+            LOG.debug("Got item page %s for URL %s", page, url)
+            page_items = len(response)
```

Comment:
This looping / pagination code is confusing. The while loop continues as long as `page_items` > `0`, but you set page_items to `0`, and to `len(response)` - which is the number of `items` retrieved in this iteration.

Reply:
The REST API doesn't return any cursor/metadata information - just a list of matching objects. If the response isn't a list (eg an error) it defaults to the page_items=0 value. It keeps looping through getting the next `per_page` items until it gets to 0. It could just test for `page_items < per_page` since that should identify the last page reliably.

I should add some better error handling. I've done some maths on the concurrency so we shouldn't hit the rate limits.

Reply:
I've raised an issue already about this function.